_posts/2018/2018-09-02-careful-with-present-when-using-activerecord-scopes.md

```diff
@@ -0,0 +1,61 @@
+---
+title: Careful with present? when using ActiveRecord scopes
+description: 'Mostly it’s fine, until your app goes boom.'
+layout: article
+category: ruby
+image:
+  base: '2018/careful-with-present-when-using-activerecord-scopes'
+  alt: 'Stadtbibliothek, Stuttgart, Deutschland'
+  source: 'https://unsplash.com/photos/PkbZahEG2Ng'
+  credit: 'Tobias Fischer'
+---
+
+When using Rails, we tend to use `#present?` in many places as it is added to _all_ Ruby objects by Rails. For example, you might use it in a view to check if an array has some elements before trying to loop through it and display some data.
+
+However using `#present?` on models with large datasets can create unforeseen problems.
+
+
+## Instead of…
+
+...using `#present?` on an Active Relation (a scope on an Active Record class)
+
+```ruby
+books = Book.recently_released
+if books.present?
+  books.paperbacks.each { |book| puts book.title }
+end
+```
+
+
+## Use…
+
+...`#any?` or `#exists?`
+
+```ruby
+books = Book.recently_released
+if books.any?
+  books.paperbacks.each { |book| puts book.title }
+end
+```
+
+
+## But why?
+
+When you call `#present?` it instantiates the objects into an array in memory before checking whether that array is empty or not. In most cases this will be fine but in extreme cases, say complex models or when using a large dataset, you might see very high memory usage and slow performing requests.
```

Comment:
This needs a bit more clarification, since, if the 'array' is empty then no objects will be instantiated, right?

So in the example you give, if there are no `recently_released` books, what is instantiated when you call `#present?` on `books` — is it all the objects in `Book`?

Reply:
It's more that if there are _lots_ and then you only use some (or don't use them at all).